compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,517 @@
+        if (
+          lvalue.identifier.name == null &&
+          value.place.identifier.name !== null &&
+          !usedOutside
+        ) {
+          temporaries.set(lvalue.identifier, value.place.identifier);
+        }
+      }
+    }
+    propertyLoadInfo.set(block.id, propertyLoads);
+  }
+  return {temporaries, properties, propertyLoadInfo};
+}
+
+function collectNodes(
+  fn: HIRFunction,
+  sidemap: TemporariesSidemap,
+): Map<BlockId, BlockInfo> {
+  /**
+   * Due to current limitations of mutable range inference, there are edge cases in
+   * which we infer known-immutable values (e.g. props or hook params) to have a
+   * mutable range and scope.
+   * (see `destructure-array-declaration-to-context-var` fixture)
+   * We track known immutable identifiers to reduce regressions (as PropagateScopeDeps
+   * is being rewritten to HIR).
+   */
+  const knownImmutableIdentifiers = new Set<Identifier>();
+  if (fn.fnType === 'Component' || fn.fnType === 'Hook') {
+    for (const p of fn.params) {
+      if (p.kind === 'Identifier') {
+        knownImmutableIdentifiers.add(p.identifier);
+      }
+    }
+  }
+  const tree = new Tree();
+  const nodes = new Map<BlockId, BlockInfo>();
+  for (const [_, block] of fn.body.blocks) {
+    const propertyLoadInfos = sidemap.propertyLoadInfo.get(block.id);
+    CompilerError.invariant(propertyLoadInfos != null, {
+      reason: '[CollectHoistablePropertyLoads] block info missing',
+      loc: GeneratedSource,
+    });
+
+    const assumedNonNullObjects = new Set<PropertyLoadNode>();
+    for (const {instr, property} of propertyLoadInfos) {
+      const object = instr.value.object.identifier;
+      const propertyNode = tree.getPropertyLoadNode(property);
+      /**
+       * Since this runs *after* buildReactiveScopeTerminals, identifier mutable ranges
+       * are not valid with respect to current instruction id numbering.
+       * We use attached reactive scope ranges as a proxy for mutable range, but this
+       * is an overestimate as (1) scope ranges merge and align to form valid program
+       * blocks and (2) passes like MemoizeFbtAndMacroOperands may assign scopes to
+       * non-mutable identifiers.
+       *
+       * See comment at top of function for why we track known immutable identifiers.
+       */
+      const isMutableAtInstr =
+        object.mutableRange.end > object.mutableRange.start + 1 &&
+        object.scope != null &&
+        inRange(instr, object.scope.range);
+      if (
+        !isMutableAtInstr ||
+        knownImmutableIdentifiers.has(propertyNode.fullPath.identifier)
+      ) {
+        let curr = propertyNode.parent;
+        while (curr != null) {
+          assumedNonNullObjects.add(curr);
+          curr = curr.parent;
+        }
+      }
+    }
+
+    nodes.set(block.id, {
+      block,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
```

Comment:
> Gotcha. That makes sense. I suspect this can still be simplified - in particular, i had to remap the variable names in my head to help keep track of things. See comments for the most unclear cases.
> 
> One thing that strikes me about this algorithm is that it isn't accurate in the sense that it doesn't account for the actual live range of values. During the initial collectNodes stage we account for mutable ranges when deciding if a property access can locally be considered to be non-nullable. But then in the deriveNonNull phase we don't have access to the mutable ranges anymore. AFAICT, if you have:
> 
```
bb0:
  let x = null
  if cond then bb1 else bb2
bb1:
  x = {}
  goto bb3
bb2:
  ...
  goto bb3:
bb3:
  x.foo.bar
```
> 
> this algorithm will infer `x.foo.bar` as non-nullable for bb0, via the intersection of bb1 & bb2 which in turn comes from bb3. So the results are actually inaccurate, but that's okay because of how we filter out dependencies in PropagateScopeDeps. 
> 
> This isn't necessarily a blocker, but it isn't great to have a pass that returns results that are technically incorrect — it's very easy to accidentally use them in a different way later, w/o realizing that important context. Thoughts on this, and whether we could make the results guaranteed accurate?

([comment from @josephsavona](https://github.com/facebook/react/pull/30079#discussion_r1675380685))

I added a comment at the top of file (on the exported function) to note this technical inaccuracy. I'm inclined to leave it for now, as reusing this logic comes with other potential gotchas (e.g. use of an identifier outside of block scope, use after reassignment, etc)